### PR TITLE
bootstrap: clear miri's ui test deps when rustc changes

### PR DESCRIPTION
**Pull request by RalfJung**
@onur-ozkan I hope this makes sense... it's quite hard to test this, but it did at least cause one rebuild of the right files for me, and then it doesn't seem to keep rebuilding, so at first sight this look at least better than the status quo.

### CODE REVIEW

File: src/bootstrap/src/core/build_steps/test.rs [hunk 1]
```diff
@@ -520,6 +520,14 @@ impl Step for Miri {
         builder.ensure(compile::Std::new(target_compiler, host));
         let host_sysroot = builder.sysroot(target_compiler);
 
+        // Miri has its own "target dir" for ui test dependencies. Make sure it gets cleared
+        // properly when rustc changes. Similar to `Builder::cargo`, we skip this in dry runs to
+        // make sure the relevant compiler has been set up properly.
+        if !builder.config.dry_run() {
+            let ui_test_dep_dir = builder.stage_out(host_compiler, Mode::ToolStd).join("ui");
+            builder.clear_if_dirty(&ui_test_dep_dir, &builder.rustc(host_compiler));
```
Comment (onur-ozkan) [inline src/bootstrap/src/core/build_steps/test.rs:528]: What is that "target dir" called? It feels like this is going to affect other tests; it doesn't seem miri-specific logic.

Comment (RalfJung) [reply src/bootstrap/src/core/build_steps/test.rs:528]: This ends up being `build/host/stage0-tools/ui` on my system. It is true that this is not Miri-specific. There is a `miri` subfolder inside this that we could clear, but strangely there is also `build/host/stage0-tools/ui/tests` with a whole bunch of empty directories (one for each directory in the Miri test suite), so Miri's artifacts are not contained to the `miri` subfolder.

AFAIK ui_test picks its output dir inside the "target dir" of the project it is being built for, so when a build environment puts many projects into the same "target dir" things can start becoming confusing... the logic for that is [here](https://github.com/oli-obk/ui_test/blob/b108b5fe1d35acb4c818b521b55bd71416aa42fb/src/config.rs#L144-L147).
Cc @oli-obk

**Comment by RalfJung** [comment]
https://github.com/rust-lang/miri/pull/3790 makes us use a Miri-specific folder.

**Comment by RalfJung** [comment]
I'm afraid only clearing when rustc changes isn't enough... we need to also clear whenever the Miri sysroot gets rebuilt. Cargo doesn't seem to recognize that the standard library changes, and then we still get build failures as before.

I don't know a good way to implement that, though...

**Comment by onur-ozkan** [comment]
We could clear it unconditionally in a similar way to how we handle docs:
https://github.com/rust-lang/rust/blob/730d5d4095a264ef5f7c0a0781eea68c15431d45/src/bootstrap/src/core/builder.rs#L1495-L1510

**Pull request merged**